frontend-wasm/src/ssa.rs

```diff
@@ -0,0 +1,552 @@
+    results: Vec<Value>,
+
+    /// Side effects accumulated in the `use_var`/`predecessors_lookup` state machine.
+    side_effects: SideEffects,
+
+    /// Reused storage for cycle-detection.
+    visited: EntitySet<Block>,
+
+    /// Storage for pending variable definitions.
+    variable_pool: ListPool<Variable>,
+
+    /// Storage for predecessor definitions.
+    inst_pool: ListPool<Inst>,
+}
+
+/// An opaque reference to a mutable variable.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Variable(u32);
+entity_impl!(Variable, "var");
+
+/// Side effects of a `use_var` or a `seal_block` method call.
+#[derive(Default)]
+pub struct SideEffects {
+    /// When a variable is used but has never been defined before (this happens in the case of
+    /// unreachable code), a placeholder `iconst` or `fconst` value is added to the right `Block`.
+    /// This field signals if it is the case and return the `Block` to which the initialization has
+    /// been added.
+    pub instructions_added_to_blocks: Vec<Block>,
+}
+
+impl SideEffects {
+    fn is_empty(&self) -> bool {
+        self.instructions_added_to_blocks.is_empty()
+    }
+}
+
+#[derive(Clone)]
+enum Sealed {
+    No {
+        // List of current Block arguments for which an earlier def has not been found yet.
+        undef_variables: EntityList<Variable>,
+    },
+    Yes,
+}
+
+impl Default for Sealed {
+    fn default() -> Self {
+        Sealed::No {
+            undef_variables: EntityList::new(),
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+struct SSABlockData {
+    // The predecessors of the Block with the block and branch instruction.
+    predecessors: EntityList<Inst>,
+    // A block is sealed if all of its predecessors have been declared.
+    sealed: Sealed,
+    // If this block is sealed and it has exactly one predecessor, this is that predecessor.
+    single_predecessor: PackedOption<Block>,
+}
+
+impl SSABuilder {
+    /// Clears a `SSABuilder` from all its data, letting it in a pristine state without
+    /// deallocating memory.
+    pub fn clear(&mut self) {
+        self.variables.clear();
+        self.ssa_blocks.clear();
+        self.variable_pool.clear();
+        self.inst_pool.clear();
+        debug_assert!(self.calls.is_empty());
+        debug_assert!(self.results.is_empty());
+        debug_assert!(self.side_effects.is_empty());
+    }
+
+    /// Tests whether an `SSABuilder` is in a cleared state.
+    pub fn is_empty(&self) -> bool {
+        self.variables.is_empty()
+            && self.ssa_blocks.is_empty()
+            && self.calls.is_empty()
+            && self.results.is_empty()
+            && self.side_effects.is_empty()
+    }
+}
+
+/// States for the `use_var`/`predecessors_lookup` state machine.
+enum Call {
+    UseVar(Inst),
+    FinishPredecessorsLookup(Value, Block),
+}
+
+/// Emit instructions to produce a zero value in the given type.
+fn emit_zero(_ty: &Type, //, mut cur: FuncCursor
```

Comment:
It's not exactly the same, `translation_utils::emit_zero` inserts at the last position but here we want to insert at the start of the block. It is needed to handle the case when the variable is used, but we cannot find its definition in the predecessors. So, we want to initialize the variable (rather than failing) at the start of the block. We need some sort of function cursor to set an arbitrary insertion point in addition to the builder interface. I could not find one. I would use it in `translation_utils::emit_zero` and call it here.

Reply:
Ah, I see. The `DataFlowGraph` provides `insert_insert` as a primitive, which allows you to specify an `InsertionPoint`, which can be either the start/end of a `Block`, or before/after an `Inst`, but I haven't exposed that in a very convenient way in the `DefaultInstBuilder`. All of the `InstBuilder` implementations actually use this under the covers, so its really just a matter of surfacing it better in one of the builders.

Reply:
Got it. I created https://github.com/0xPolygonMiden/miden-ir/issues/40